[PT Run] Registry plugin

TobiasSekan opened the pull request:
## Summary of the Pull Request
Plugin for PT run to find Registry keys of Windows 🎉 🥳 

## A few pictures

Show all registry main keys, when start with `:`
![2021-01-08 13_38_08-](https://user-images.githubusercontent.com/17874713/104016565-ddae0880-51b6-11eb-8e17-1500e18cc1fd.png)

Show a filtered list, when the match found more than one key
![2021-01-08 13_33_05-](https://user-images.githubusercontent.com/17874713/104016734-2d8ccf80-51b7-11eb-9774-727a41975e08.png)



Show all values from a registry key, when search query end with `\\` and only one registry key match the search
![2021-01-08 13_24_46-](https://user-images.githubusercontent.com/17874713/104016560-dab31800-51b6-11eb-99ab-951bd290e0d1.png)

## Features
* support search for full main keys and shortcuts (e.g. `HKLM`, `HKCU`)
* show all root keys when a user type only `:` `:HKEY` or `:HKEY_`
* match keys they start with shortcut + search
* Show result of found keys (count of sub-keys, count of values)
* case-intensive
* command: open direct in `regedit.exe`
* command: copy key/name/value to clipboard
* search inside a registry for value names and values

## PR Checklist
* [x] Applies to #5495
* [x] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/PowerToys) and sign the CLA
- [x] **Tests:** Added/updated and all pass
- [ ] **Installer:** Added/updated and all pass
- [x] **Localization:** All end user facing strings can be localized
- [x] **Docs:** Added/ updated
- [x] **Binaries:** Any new files are added to WXS / YML
   - [ ] No new binaries
   - [x] [YML for signing](https://github.com/microsoft/PowerToys/blob/master/.pipelines/pipeline.user.windows.yml#L68) for new binaries
   - [ ] [WXS for installer](https://github.com/microsoft/PowerToys/blob/master/installer/PowerToysSetup/Product.wxs) for new binaries

## Info on Pull Request

* New plugin
* Developer documentation
* UnitTests

## Validation Steps Performed

manual test and unit tests

alannt777 commented:
**Suggestion**: New icon for registry keys?

htcfreek commented:
> **Suggestion**: New icon for registry keys?

cc: @niels9001

htcfreek commented:
suggestion:
I the user types a : after the key the values within the key are shown.

`HKLM:`
```
<Default>=
(Type: REG_SZ)


Value1=0x000001
(Type: REG_DWORD)
```

niels9001 commented:
@TobiasSekan 

Here's a light and dark theme icon that might be suitable. It somewhat represents the regedit.exe app icon.

![registry light](https://user-images.githubusercontent.com/9866362/98598130-4d4c6800-22da-11eb-834e-7de3696b1212.png)
![registry dark](https://user-images.githubusercontent.com/9866362/98598135-4de4fe80-22da-11eb-8922-6cded19b8fa2.png)

jsoref commented on src/modules/launcher/Plugins/Microsoft.Plugin.Registry/Main.cs:
```diff
@@ -0,0 +1,127 @@
+                    catch (System.ComponentModel.Win32Exception)
+                    {
+                        MessageBox.Show(
+                            "You have not enought rights to open the Registry Editor",
```
`enough`

jsoref replied:
Also, this is grammatically awkward, it should be `You do not have ...`

jsoref commented on src/modules/launcher/Plugins/Microsoft.Plugin.Registry/Helper/RegistryHelper.cs:
```diff
@@ -0,0 +1,175 @@
+        };
+
+        /// <summary>
+        /// Try to find a registery key based on the given serach query
```
`registry` (and elsewhere)

jsoref replied:
`serach`

jsoref commented on src/modules/launcher/Plugins/Microsoft.Plugin.Registry/Helper/RegistryHelper.cs:
```diff
@@ -0,0 +1,175 @@
+
+        internal static void OpenRegisteryKey(string fullKey)
+        {
+            // it's impossible to direct open a key via command-line option, so we must override the last remembert key
```
`remember`

alannt777 replied:
`OpenRegistryKey`

jsoref commented on src/modules/launcher/Plugins/Microsoft.Plugin.Registry/Helper/RegistryHelper.cs:
```diff
@@ -0,0 +1,175 @@
+
+        internal static ICollection<(string, RegistryKey?, Exception?)> SerachForKey(in RegistryKey mainKey, in string subKeyPath)
+        {
+            Debug.WriteLine($"Serarch for {mainKey.Name}\\{subKeyPath}\n");
```
`Search`

jsoref commented on src/modules/launcher/Plugins/Microsoft.Plugin.Registry/plugin.json:
```diff
@@ -0,0 +1,12 @@
+{
+  "ID": "303417D927BF4C97BCFFC78A123BE0C8",
```
You're going to want to add an entry to:: https://github.com/TobiasSekan/PowerToys/blob/RegistryPlugin/.github/actions/spell-check/patterns.txt for this one, e.g.:
```
# plugin.json
^  "ID": "[0-9A-F]{32}",$
```

jsoref reviewed (changes requested):
Note that the bot leaves fairly detailed tags identifying what it doesn't like: https://github.com/TobiasSekan/PowerToys/commit/ab7ac24f84f4b062b804222a0414b30b0d1427c1 (but only for the first 10).
The rest can be found in the Action log: https://github.com/TobiasSekan/PowerToys/runs/1374592156?check_suite_focus=true#step:3:41

TobiasSekan commented:
@htcfreek 

Something like this?

![2020-11-10 12_23_47-](https://user-images.githubusercontent.com/17874713/98668220-dd78c480-234f-11eb-87aa-95ddafe7c8f5.png)

TobiasSekan commented:
> 
> 
> @TobiasSekan
> 
> Here's a light and dark theme icon that might be suitable. It somewhat represents the regedit.exe app icon.
> 
> ![registry light](https://user-images.githubusercontent.com/9866362/98598130-4d4c6800-22da-11eb-834e-7de3696b1212.png)
> ![registry dark](https://user-images.githubusercontent.com/9866362/98598135-4de4fe80-22da-11eb-8922-6cded19b8fa2.png)

Thank you 👍

htcfreek commented:
> @htcfreek 
> 
> Something like this?
> 
> ![2020-11-10 12_23_47-](https://user-images.githubusercontent.com/17874713/98668220-dd78c480-234f-11eb-87aa-95ddafe7c8f5.png)
> 

**Feedback for sub text:**

But instead of `Value (Type)` I would write `Type: Reg_SZ * Value: abcdefg` in the sub text to be in line with the sub text for keys and to see the type also with long values.

Is it possible to end long cutted values with ` ... `?

Maybe it make sense to show a text like `See more on mouse over` instead showing the value in the sub text.

htcfreek commented:
@TobiasSekan 
My idea with the `:` was great I think. But we have a Problem: What will we do if the key path contains a `:`?

TobiasSekan commented:
Only the `backslash(\)` isn't allowed in a key name

![2020-11-11 14_48_50-Registry Editor](https://user-images.githubusercontent.com/17874713/98819384-13d84180-242d-11eb-9dab-5aaf1a48245e.png)

All other signs are okay

![2020-11-11 14_47_44-Registry Editor](https://user-images.githubusercontent.com/17874713/98819392-189cf580-242d-11eb-9e62-acf4d7124824.png)

TobiasSekan commented:
>     1. Do you count the default value too?

I use the WIN32 API and they don't count the default value.

>     2. Does the plugin needs admin rights and what happened, if you navigate into a permitted key?

Admin rights are only need for protected keys and to start `regedit.exe`

None admin user, typical get this:
![2020-11-11 15_06_27-PowerLauncher](https://user-images.githubusercontent.com/17874713/98821236-9104b600-242f-11eb-8b33-5a2004c65570.png)

htcfreek commented:
> >     1. Do you count the default value too?
> 
> I use the WIN32 API and they don't count the default value.
> 
> >     2. Does the plugin needs admin rights and what happened, if you navigate into a permitted key?
> 
> Admin rights are only need for protected keys and to start `regedit.exe`
> 
> None admin user, typical get this:
> ![2020-11-11 15_06_27-PowerLauncher](https://user-images.githubusercontent.com/17874713/98821236-9104b600-242f-11eb-8b33-5a2004c65570.png)
> 

@TobiasSekan 
Maybe we can combine it with an keyword which is entered by pressing enter:
- subkey
- values

htcfreek commented:
> Only the `backslash(\)` isn't allowed in a key name
> 
> ![2020-11-11 14_48_50-Registry Editor](https://user-images.githubusercontent.com/17874713/98819384-13d84180-242d-11eb-9dab-5aaf1a48245e.png)
> 
> All other signs are okay
> 
> ![2020-11-11 14_47_44-Registry Editor](https://user-images.githubusercontent.com/17874713/98819392-189cf580-242d-11eb-9e62-acf4d7124824.png)
> 
> So, i will add your proposal
> >     * `Enter` to show the values within a key and `Strg + Backspace` to switch from the value mode in the key mode.
> 

@TobiasSekan 
Maybe we can combine it with an keyword which is entered by pressing enter:
- subkey
- values

Example: http://www.wox.one/plugin/129

htcfreek commented:
@TobiasSekan 
We should add an extended open command (`Ctrl+Shift+Enter`) to open in Regedit as admin (show uac).
Not sure if this works with predefined open in destination. 🤔

alannt777 commented on src/modules/launcher/Plugins/Microsoft.Plugin.Registry/Helper/ContextMenuHelper.cs:
```diff
@@ -0,0 +1,68 @@
+            }
+            catch (Exception exception)
+            {
+                Log.Exception("Error on open Windows registry editor", exception, typeof(Main));
```
"opening the Windows reg..."?

alannt777 commented on src/modules/launcher/Plugins/Microsoft.Plugin.Registry/Helper/RegistryHelper.cs:
```diff
@@ -0,0 +1,214 @@
+        /// <param name="fullKey">The registry key to open</param>
+        internal static void OpenRegistryKey(in string fullKey)
+        {
+            // it's impossible to direct open a key via command-line option, so we must override the last remember key
```
directly ;)

TobiasSekan commented:
@htcfreek 
Hello, i have try to use `IResultUpdated` but, it don't refresh the surface (the list is still refreshed, but i think there is a missing `Dispatcher` call to refresh the surface or similar)

So as current workaround you can write two backslashes at the end of the search query
![2020-11-23 13_32_54-PowerLauncher](https://user-images.githubusercontent.com/17874713/99963566-43267100-2d92-11eb-957d-3a534ed93da0.png)

TobiasSekan commented:
Should I at underscores in the values for easier reading ???
```
// without underscores (current behavior)
0x00DA3700 (14300928)

// with underscores
0x_00_DA_37_00 (14_300_928)
```

TobiasSekan commented:
My idea for the very long keys
1. Try to use a short main key name
2. Or three left dots

![2020-11-23 16_52_42-PowerLauncher](https://user-images.githubusercontent.com/17874713/99983975-d53b7300-2dac-11eb-980b-6e9b95aaf0d0.png)

htcfreek commented:
> My idea for the very long keys
> 1. Try to use a short main key name
> 2. Or three left dots
> 
> ![2020-11-23 16_52_42-PowerLauncher](https://user-images.githubusercontent.com/17874713/99983975-d53b7300-2dac-11eb-980b-6e9b95aaf0d0.png)
> 

Like the idea.

The short version of the main key seems to be incorrect. I would expect `HKCU` instead of `HCCU`.

htcfreek commented:
> @htcfreek 
> Hello, i have try to use `IResultUpdated` but, it don't refresh the surface (the list is still refreshed, but i think there is a missing `Dispatcher` call to refresh the surface or similar)
> 
> So as current workaround you can write two backslashes at the end of the search query
> ![2020-11-23 13_32_54-PowerLauncher](https://user-images.githubusercontent.com/17874713/99963566-43267100-2d92-11eb-957d-3a534ed93da0.png)

@TobiasSekan 
If it's not possible to start a value name with a blank ` ` then maybe we can allow to filter the values after typing `\\ `.

TobiasSekan commented:
@htcfreek 
What you think about that?
You can direct type the value name after the two backslahes
![2020-11-26 11_59_27-PowerToys (Running) - Microsoft Visual Studio Preview](https://user-images.githubusercontent.com/17874713/100344515-38ade680-2fe1-11eb-8822-2b45d7d0c2de.png)

Should I expand the search so that values can be search too?

htcfreek commented:
> @htcfreek 
> What you think about that?
> You can direct type the value name after the two backslahes
> ![2020-11-26 11_59_27-PowerToys (Running) - Microsoft Visual Studio Preview](https://user-images.githubusercontent.com/17874713/100344515-38ade680-2fe1-11eb-8822-2b45d7d0c2de.png)

We should expect a blank between key and name because a value can start with a `\`.

> Should I expand the search so that values can be search too?
> 

Not sure if this is a good idea. Could lead to many results.

TobiasSekan commented:
We have a problem 🙅  `query.Search` and `query.RawQuery` always trim the last spaces
(see `PowerLauncher.Plugin.QueryBuilder`)

so I can find out that a user only write `\ `  (backslash with space) to show all values of a key

alannt777 commented on src/modules/launcher/Plugins/Microsoft.Plugin.Registry/plugin.json:
```diff
@@ -0,0 +1,12 @@
+{
+  "ID": "303417D927BF4C97BCFFC78A123BE0C8",
+  "ActionKeyword": "*",
```
Is "*" an action keyword or is it just a placeholder?

TobiasSekan replied:
It's gives all queries to the plugin. Without any filtering.

TobiasSekan replied:
I have to try use `ActionKeywords: [ "HK", "hk", "Hk", "hK" ]`, but WOX always print the keyword + search in the ComboBox.
![2020-11-27 09_24_35-PowerLauncher](https://user-images.githubusercontent.com/17874713/100427397-aa923880-3092-11eb-82d2-8227f6807a5c.png)
It is not possible to override this behavior. Currently I always override the `result.QueryTextDisplay` to catch short base keys.

htcfreek commented:
> We have a problem 🙅  `query.Search` and `query.RawQuery` always trim the last spaces
> (see `PowerLauncher.Plugin.QueryBuilder`)
> 
> so I can find out that a user only write `\ `  (backslash with space) to show all values of a key

Don't understand.
What would you expect to find out?

TobiasSekan commented:
Hello,

before the last change when a user type two backslashes `\\` and nothing more, I show all value names of the current selected key.

Now I looking for a backslash and a space and the characters after the space. When a user write `hklm\cons\ color`, I search for first key they start with `cons` in the base key `HKEY_LOCAL_MACHINE` and filter all value names of this key that contain `color`, so far so good.

But, i need a simple concept to show all value names of a key and currently.

And currently I can't figure out that a user have write `hklm\cons\ ` (end with space). I get always only `hklm\cons\` (without a space) and so I show all keys behind the given query key of `hklm\cons\` and not the value inside the key.

htcfreek commented:
> Hello,
> 
> before the last change when a user type two backslashes `\\` and nothing more, I show all value names of the current selected key.
> 
> Now I looking for a backslash and a space and the characters after the space. When a user write `hklm\cons\ color`, I search for first key they start with `cons` in the base key `HKEY_LOCAL_MACHINE` and filter all value names of this key that contain `color`, so far so good.
> 
> But, i need a simple concept to show all value names of a key.
> 
> And currently I can't figure out that a user have write `hklm\cons\ ` (end with space). I get always only `hklm\cons\` (without a space) and so I show all keys behind the given query key of `hklm\cons\` and not the value inside the key.
> 
> 

Can we lock at indexer plugin, how it's done there.

TobiasSekan commented:
The `Indexer Plugin` don't have this problem.
It show all (folders and files) after the backslash or filter the results by a keyword after the backslash

Possible solutions:

1. Introduce extra action keywords (`>` and `*`) similar to
![2020-11-27 08_58_30-PowerLauncher](https://user-images.githubusercontent.com/17874713/100425104-e4614000-308e-11eb-9cc5-f961c840318d.png)
* Pro: No new keywords
* Cons: User must know these keywords
2. Show all sub-keys and value names after the `\`
 * Pro: Same behavior as the file explorer and the index plugin
 * Cons: Not the same behavior as the registry editor
3. Introduce a new property in `Query.cs` that contains the real untouched input
 * Pro: No side effects
 * Cons: Extra property
4. Use a placeholder to show all values of a key like `*`
 * Pro: Use of the typical `Show all` placeholder
 * Cons: A value name can start with have a `*`
 * Cons: Introduce extra syntax that are hard to remove later (When people have got used to it)

I prefer a combination of  1 and 2

alannt777 commented on doc/devdocs/modules/launcher/plugins/registry.md:
```diff
@@ -0,0 +1,96 @@
+# Registry Plugin
+
+The registry plugin allow the users to search into the Windows registry.
```
nit: "allows"

alannt777 replied:
"The registry plugin allow**s**  users to **search the** Windows registry." sounds better

alannt777 commented on doc/devdocs/modules/launcher/plugins/registry.md:
```diff
@@ -0,0 +1,96 @@
+
+*Note: Linux and macOS program ports typical store the settings in it's own configuration files and not in the Windows registry.*
+
+For more information about the Windows registry, see [the official documentation about it](https://docs.microsoft.com/en-us/windows/win32/sysinfo/registry).
```
nit: "the official documentation about it" seems too verbose. I think "the official documentation" is fine

TobiasSekan commented:
Ping @htcfreek -> https://github.com/microsoft/PowerToys/pull/7951#issuecomment-734701890

htcfreek commented:
@TobiasSekan 
Sorry that I have missed to answer.

I think with 1. we could get problems with the allowed characters in value names and we should only have one keyword.

Regarding 2: This could lead to to many results. If we do this we should have different icons for keys and values. Maybe for each value type.

I would prefer solution 3: `\` shows all subkeys and `\\` shows all values in a key. Writing a text behind the `\\` gets interpreted as filter for the value name.

TobiasSekan commented:
@jsoref How can i fix this?

>New misspellings found, please review:
>    BCFFC

jsoref commented on .github/actions/spell-check/expect.txt:
```diff
@@ -2531,3 +2538,4 @@ ZONESETCHANGE
 Zoneszonabletester
 Zoomusingmagnifier
 zzz
+`303417D927BF4C97BCFFC78A123BE0C8`
```
Expect only tolerates words it recognizes, so this won't work. Drop it.

jsoref commented on doc/devdocs/modules/launcher/plugins/registry.md:
```diff
@@ -0,0 +1,97 @@
+| --------------- | ---------------------------------- |
+| ActionKeyword   | `*`                                |
+| ExecuteFileName | `Microsoft.Plugin.Registry.dll`    |
+| ID              | `303417D927BF4C97BCFFC78A123BE0C8` |
```
Add a pattern for this. It'll be similar to the other pattern, but it's why the checker isn't happy.

```
^\s+\|\s+ID\s+\|\s*\`[0-9A-F]{32}\`
```

htcfreek commented:
@TobiasSekan 
Are I am right that we don't need a action key because the suffix for the hive is working as action key?

TobiasSekan commented:
> 
> 
> @TobiasSekan
> Are I am right that we don't need a action key because the suffix for the hive is working as action key?

This plugin currently allow another `Registry Hive`, when anybody need this I can make this on a new Issue.

htcfreek commented:
> > 
> > 
> > @TobiasSekan
> > Are I am right that we don't need a action key because the suffix for the hive is working as action key?
> 
> This plugin currently allow no other `Registry Hive`, when anybody need this I can make this on a new Issue.

Hm. I think you misunderstood me. For other plugins (e. g. Service plugin) we have set an action key. So their results will only shown when you type this character in front of your query. Do we need this for the registry plugin or does the plugin only react if the first word is a reg hive.

jsoref commented:
@TobiasSekan, the bot tries to highlight where it finds misspellings, e.g. of `name`: https://github.com/TobiasSekan/PowerToys/commit/0275ac434783c89629fb5f8a2cc8f8d25f8face4#diff-bb4d059fad2b893abe18abff0f4add5f7d556303b9994f98085eef96131464a3R25
- I'd expect GitHub notifications to tell you about these (I generally get email for them)

It looks like the regular expression I suggested didn't work. 

You can experiment using:
https://github.com/check-spelling/check-spelling/wiki/Looking-for-items-locally (I'm still evolving this, eventually if it's helpful, the suggestion might include a reference to it to help people debug.)

To speed things up, you can set `$files` to be just the two files that have `BCFFC` (the delimiter for the variable is `\n`, or you could do it one file at a time), and try using `peek BCFFC`.

I might be able to look into this expression today. If not, maybe on Sunday.

TobiasSekan commented:
> 
> 
> > > @TobiasSekan
> > > Are I am right that we don't need a action key because the suffix for the hive is working as action key?
> > 
> > 
> > This plugin currently allow no other `Registry Hive`, when anybody need this I can make this on a new Issue.
> 
> Hm. I think you misunderstood me. For other plugins (e. g. Service plugin) we have set an action key. So their results will only shown when you type this character in front of your query. Do we need this for the registry plugin or does the plugin only react if the first word is a reg hive.

Hello, ah okay

Currently the plugin ignore all request that have less then two characters and the query must start with `HKEY` (case-intensive) or a short-hand e.g. `HKCU` (case-intensive)


I can add a action key, but which?

htcfreek commented:
> > 
> > 
> > > > @TobiasSekan
> > > > Are I am right that we don't need a action key because the suffix for the hive is working as action key?
> > > 
> > > 
> > > This plugin currently allow no other `Registry Hive`, when anybody need this I can make this on a new Issue.
> > 
> > Hm. I think you misunderstood me. For other plugins (e. g. Service plugin) we have set an action key. So their results will only shown when you type this character in front of your query. Do we need this for the registry plugin or does the plugin only react if the first word is a reg hive.
> 
> Hello, ah okay
> 
> Currently the plugin ignore all request that have less then two characters and the query must start with `HKEY` (case-intensive) or a short-hand e.g. `HKCU` (case-intensive)
> 
> 
> I can add a action key, but which?
> 
> 

Because the query must start with the main hive name I think we don't need an action key.

jsoref commented on .github/actions/spell-check/patterns.txt:
```diff
@@ -32,3 +32,12 @@ TestCase\("[^"]+"
+\[DataRow\(.*\)\]
+
+# Id info inside markdown file (registry.md) 
+^\s+\|\s+ID\s+\|\s*\`[0-9A-F]{32}\`
```
Sorry, turns out there was _no_ whitespace at the beginning of the line (I wrote that instruction from a phone and guessed wrong about the markup):
```suggestion
^\|\s+ID\s+\|\s*\`[0-9A-F]{32}\`
```

jsoref commented on src/modules/launcher/Plugins/Microsoft.Plugin.Registry/Helper/ContextMenuHelper.cs:
```diff
@@ -0,0 +1,140 @@
+        /// Return a list with all context menu entries for the given <see cref="Result"/>
+        /// </summary>
+        /// <param name="result">The result for the context menu entires</param>
+        /// <param name="assemblyName">The anme of the this assembly</param>
```
```suggestion
        /// <param name="assemblyName">The name of the this assembly</param>
```

jsoref commented on src/modules/launcher/Plugins/Microsoft.Plugin.Registry/Main.cs:
```diff
@@ -0,0 +1,171 @@
+            var (baseKeyList, subKey) = RegistryHelper.GetRegistryBaseKey(queryKey);
+            if (baseKeyList is null)
+            {
+                // no base key found -> test is the key starts with "HKE" or "HKEY" (case-insensitive)
```
If `"HKE"` is normally a misspelling, I'd probably add:

jsoref replied:
If `HKE` is normally a misspelling, I'd suggest adding:
```
"HKE" or "HKEY"
```
to `patterns.txt` instead of adding `HKE` to `expect.txt`

jsoref commented on src/modules/launcher/Plugins/Microsoft.Plugin.Registry.UnitTest/Helper/RegistryHelperTest.cs:
```diff
@@ -0,0 +1,73 @@
+        [TestMethod]
+        public void GetRegistryBaseKeyTestMoreThanOneBaseKey()
+        {
+            var (baseKeyList, _) = RegistryHelper.GetRegistryBaseKey("HKC\\Control Panel\\Accessibility");
```
Same sort of thing wrt `HKC` ... although, maybe just add a `/* #no-spell-check-line */` to the end of the line and then add:
```
^.*/\* #no-spell-check-line \*/.*$
```
to `patterns.txt`
--- I haven't really standardized on a marker for this purpose, but this is the example pattern in https://github.com/check-spelling/check-spelling/wiki/Configuration-Examples:-patterns

`expect.txt` is a compromise between things which are always correct (which technically can live in `allow.txt`) and things which are transient, but which hopefully aren't easy typos for more correct things. For things that are exceptionally acceptable in specific instances, it's best to use `patterns.txt`, otherwise, in other cases where they're typos, the presence in `expect.txt` will result in the typo forms sneaking in.

htcfreek commented:
Do wr get this reafy for 0.29 or do we need more time?

TobiasSekan commented:
I think it's ready for review.

htcfreek reviewed (approved):
The behaviour we have now looks good to me.

@crutkas
Can we get this in for 0.29? What do you think?

crutkas commented on src/modules/launcher/Plugins/Microsoft.Plugin.Registry/plugin.json:
```diff
@@ -0,0 +1,12 @@
+{
+  "ID": "303417D927BF4C97BCFFC78A123BE0C8",
+  "ActionKeyword": "*",
```
This should only be invoked by a action keyword, not everything.

TobiasSekan replied:
fixed, it is now `"ActionKeywords": [ "HK", "hk", "Hk", "hK" ],` we need a four combinations, because it is case-sensitive

crutkas commented:
I defer to @enricogior, i'm on vacation next week.

Big bit of feedback is I believe this should only be invoked via keyword, not the global

htcfreek commented:
> I defer to @enricogior, i'm on vacation next week.
> 
> Big bit of feedback is I believe this should only be invoked via keyword, not the global

Not sure if you read our discussion above: The plugin gets invekoked as a global plugin but only shows results if the first word is the name of a registry hive (e. g. hklm or hkey_local_machine).

BTW: Have a nice vacation.

TobiasSekan commented:
@htcfreek I have add action keywords, the plugin is now

htcfreek commented:
> @htcfreek I have add action keywords, the plugin is now triggered by
> ```json
>   "ActionKeywords": [ "HK", "hk", "Hk", "hK" ],
> ``` 

And this is working with two characters? Doesn't know that we can use also two characters.

TobiasSekan commented:
> 
> 
> > @htcfreek I have add action keywords, the plugin is now triggered by
> > ```json
> >   "ActionKeywords": [ "HK", "hk", "Hk", "hK" ],
> > ```
> 
> And this is working with two characters? Doesn't know that we can use also two characters.

Yes, it was a little bit tricky to find a way to remove the space between the action keyword an the display text, but it works perfect on my PC.

Jay-o-Way commented:
> Should I add underscores in the values for easier reading ???

No. It's one value and they belong together. And imagine what would happen if you would copy and paste this value?

Jay-o-Way commented:
The end of the month is close. Would this PR make it into the next update?

crutkas commented on src/modules/launcher/Plugins/Microsoft.Plugin.Registry/plugin.json:
```diff
@@ -0,0 +1,12 @@
+{
+  "ID": "303417D927BF4C97BCFFC78A123BE0C8",
+  "ActionKeywords": [ "HK", "hk", "Hk", "hK" ],
```
why is the keyword HK versus maybe a symbol of some sort?    HK could be used by someone for a file name such as Hong Kong or HKD for their dollar.

How about **!** or something like that?

Jay-o-Way replied:
Why not use `hklm` or `hkcu` or one of the others?

crutkas replied:
actually, ! would be better for services.  

- %% maybe?

here are the current items in use: https://docs.microsoft.com/en-us/windows/powertoys/run#action-key

Action key | Action
-- | --
= | Calculator only. Example =2+2
? | File searching only. Example ?road to find roadmap.txt
. | Installed app searching only. Example .code to get Visual Studio Code
// | URLs only. Example //docs.microsoft.com to have your default browser go to https://docs.microsoft.com
< | Running processes only. Example <outlook to find all processes that contain outlook
> | Shell command only. Example >ping localhost to do a ping query

htcfreek replied:
For the service plugin we use _ at the moment.

TobiasSekan replied:
What you think about `\` (Backslash)

The Regedit Editor write in the address line
`Computer\HKEY_ ...`
so, when we reduce this to 
`\HKEY_ ...`
then is the `\` more intuitive

htcfreek replied:
I personally don't like `\` as action key for the reg plugin. `\` remind me on fie paths. What about using `_` or `$`? (For `_` please see my comment on #.)

TobiasSekan replied:
We should not use `$`,
IMHO a `$` looks like a environmental variable or a windows special folder (idea for a new plugin 💡)

TobiasSekan replied:
What you think about `:` or `::` so it would look more like PowerShell
`Get-Item -Path Registry::HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion ... `

htcfreek replied:
Don't like `:` as action key.

So I think we can use `%` like @crutkas suggested or `_` (if not used for services).
Or what about `&` or `#` (as synonym for a number setting like DWORD)?

enricogior replied:
Actually `:` seems a good choice.

TobiasSekan replied:
I have change it to `:`

crutkas commented:
This needs WSX inclusion and the YML for signing

crutkas commented:
This needs WSX inclusion and YML for signing.  that can either be a must fix for 0.31 new issue or included here.

enricogior commented:
@TobiasSekan
I've updated the initial comment with the new checklist we are now using for PRs, as per @crutkas previous comment.

htcfreek commented:
@TobiasSekan
The pictures in the main description shows that we use different characters to split the text items in the subtitles for keys and values: `-` and `*`
Is this fixed to use `-` for both result types and only the screenshot are outdated?

TobiasSekan commented:
> 
> 
> @TobiasSekan
> 
>     * The pictures in the main description shows that we use different characters (`-` and `*`) to split the text in the subtitles for keys and values. Is this fixed to use `-` for both result types and only the screenshot are outdated?
> 
>     * Will the default key be shown in the results if it's not empty?
> 
>     * In a picture above the values of a key are shown also when the key name/path is not written completely. I think the values of a key should be only shown if the key path is written completely in the search input field. Can we change this? What do you think about this?

1. Changed -> always use `-`

2. Yes the default value is shown, when not empty, I have fix the missing `(Default)` for the value name
![2021-01-08 11_28_12-](https://user-images.githubusercontent.com/17874713/104017099-c91e4000-51b7-11eb-859b-6434893329d7.png)

3. Changed -> Now only when only one sub-key is found  the user can see/search for a value in it.
![2021-01-08 13_33_05-](https://user-images.githubusercontent.com/17874713/104017239-0be01800-51b8-11eb-94b3-767971640c5a.png)

![2021-01-08 13_24_46-](https://user-images.githubusercontent.com/17874713/104017250-11d5f900-51b8-11eb-86b6-6ae84ffd690b.png)

TobiasSekan commented:
@enricogior I have update the

htcfreek commented:
@TobiasSekan 
Checks are failing.

enricogior commented:
@TobiasSekan 
for YAML it's not required to test it, when the binaries are added to the list that's good enough.
For WXS I'll take a look and let you know.

TobiasSekan commented:
> 
> 
> @TobiasSekan
> Checks are failing.

i have fix the wrong rebased solution file

htcfreek commented:
@TobiasSekan 
You wrote that the user can search for values in keys. Does this mean the user can type the value name after `\\`?

htcfreek commented:
In the service plugin we have an "open external app" icon for opening `services.msc`. Does it make sense to use this symbol here for opening `regedit` instead of the edit symbol?

What does the copy symbols mean? For me the different icons are not clear enough.

TobiasSekan commented:
> 
> 
> @TobiasSekan
> You wrote that the user can search for values in keys. Does this mean the user can type the value name after `\\`?

Yes

By value name

![2021-01-11 12_52_14-](https://user-images.githubusercontent.com/17874713/104179462-2f9b9c00-540c-11eb-87b2-3aa89a2e3865.png)

and by value

![2021-01-11 12_52_31-](https://user-images.githubusercontent.com/17874713/104179470-31fdf600-540c-11eb-991f-7daffb0f382f.png)

TobiasSekan commented:
> 
> 
> In the service plugin we have an "open external app" icon for opening `services.msc`. Does it make sense to use this symbol here for opening `regedit` instead of the edit symbol?
> 
> What does the copy symbols mean? For me the different icons are not clear enough.

1. No problem, i change it.

2. I have not found better symbols for copy different kind of values, but i can change it, when you want

The first is for copy value (only shown on values)
![2021-01-11 12_52_52-](https://user-images.githubusercontent.com/17874713/104179552-51951e80-540c-11eb-87b7-99d3523ec65f.png)

and second is for copy value name (on values) 
![2021-01-11 12_53_04-](https://user-images.githubusercontent.com/17874713/104179557-52c64b80-540c-11eb-80ac-fbbe8af68821.png)

or copy registry key (on keys)
![2021-01-11 12_58_06-](https://user-images.githubusercontent.com/17874713/104179791-b2245b80-540c-11eb-854b-cbca553b827e.png)

htcfreek commented:
> > 
> > 
> > In the service plugin we have an "open external app" icon for opening `services.msc`. Does it make sense to use this symbol here for opening `regedit` instead of the edit symbol?
> > 
> > What does the copy symbols mean? For me the different icons are not clear enough.
> 
> 1. No problem, i change it.
> 
> ![2021-01-11 13_11_18-](https://user-images.githubusercontent.com/17874713/104181219-a9cd2000-540e-11eb-8f5f-d9ef5a60bd23.png)

👍 Think this makes more sense.

> 2. I have not found better symbols for copy different kind of values, but i can change it, when you want
> 
> The first is for copy value (only shown on values)
> ![2021-01-11 12_52_52-](https://user-images.githubusercontent.com/17874713/104179552-51951e80-540c-11eb-87b7-99d3523ec65f.png)
> 
> and second is for copy value name (on values) 
> ![2021-01-11 12_53_04-](https://user-images.githubusercontent.com/17874713/104179557-52c64b80-540c-11eb-80ac-fbbe8af68821.png)
> 
> or copy registry key (on keys)
> ![2021-01-11 12_58_06-](https://user-images.githubusercontent.com/17874713/104179791-b2245b80-540c-11eb-854b-cbca553b827e.png)
> 
> 

Regarding the copy tooltips: They should show simply the action the button does.
1. I would remove the value/key name/data in the toolti.
2. I would optimize and simplify text like this:
- copy value data (shortcut)
- copy value name (shortcut)
- copy key path (shortcut)
(I think that we copy to clipboard is logical.)


@niels9001 
Do you have an idea for a maybe better copy value data symbol?

niels9001 commented:
Agree with @htcfreek, I'd keep the tooltips inline with other Run plug-ins and simplify it. You are copying object X, so there's not a need to describe it. I'd just say something like "Copy" - the selected item indicates if it's a value, key or x so there's not really a need to mention it again IMO. Also make sure you use lowercase for the shortcut: Copy (Ctrl+Shift+C) instead of (CTRL+SHIFT+C), like we do here:

![image](https://user-images.githubusercontent.com/9866362/104184180-57423280-5413-11eb-887c-9598320057f6.png)

I have no recommendation for a better icon.. tough one to solve :).

TobiasSekan commented:
@htcfreek @niels9001

htcfreek commented:
> @htcfreek @niels9001 
> I have changed the `copy value data ` to `Ctrl+Shift+D` (d for data) 
> You are okay with this?

What I mean is: Can we simplify the shortcuts:
- `Strg+C`: copy value nane/key path
- `Strg+D`: copy data
- `enter`: open in regedit

I don't think that we need additional modifier keys like `Alt` or `Shift`. It should be simple an intuitive.

crutkas commented:
Do we need a 3rd item?  Am i missing something?  I agree with @htcfreek here and doing 2 vs 3 items. Regedit also uses Key Name and that returns the full path.  I would personally include the phrase path as i don't personally think that is directly intuitive unless you live in RegEdit.  I would have always called it a path.

- default behavior: Clicking / enter key = open regedit hopefully deeplinked to that key
- Copy key name (path) = ctrl+c
- Copy data = ctrl+shift+c
   - **Why not `D`**: i do not think that is typically intuitive.  If you are trying to copy info, to the best of my knowledge, they tend to be variants of `ctrl+c`

Things of note and suggest we stay away from:
ctrl+shift+enter == this behavior is reserved typically for "run as admin".  suggest we stay away from that

htcfreek commented:
> Do we need a 3rd item?  Am i missing something?  I agree with @htcfreek here and doing 2 vs 3 items. Regedit also uses Key Name and that returns the full path.  I would personally include the phrase path as i don't personally think that is directly intuitive unless you live in RegEdit.  I would have always called it a path.
> 
> - default behavior: Clicking / enter key = open regedit hopefully deeplinked to that key
> - Copy key name (path) = ctrl+c
> - Copy data = ctrl+shift+c
>    - **Why not `D`**: i do not think that is typically intuitive.  If you are trying to copy info, to the best of my knowledge, they tend to be variants of `ctrl+c`
> 
> Things of note and suggest we stay away from:
> ctrl+shift+enter == this behavior is reserved typically for "run as admin".  suggest we stay away from that
> 
> 

I agree completely.

Yes the regedit opens in the path. We use the `LastOpendKey` value of `regedit.exe`.

htcfreek commented:
I think about if it's a good idea or a `to much` idea to have a copy path command on values too.
My usecase: I need to open regedit as different user and want to jump in the key of a value I have searched.

htcfreek commented:
I have seen in the code that we run regedit as admin. Not sure if this is a good idea. If the user has no admin rights or the uac is disabled (like in my company) it won't work. And additional I think we could get an unwanted behaviour if the path is under `hkcu` and the user has to choose a different admin account in uac because their user has only minimal rights.

TobiasSekan commented:
Address feedback:
* click/enter key = open regedit
* Copy value name / Copy key path = ctrl+c
* Copy value data = ctrl+shift+c 

* Key name is copy as `Path + \\ + KeyName `

@htcfreek
When I remove the run as admin, I can't longer open `regedit.exe` on my computer with only one user admin account

![2021-01-11 21_02_41-Error on open Registry Editor](https://user-images.githubusercontent.com/17874713/104232527-c12cfd00-5450-11eb-8bd7-f83f8b082ed7.png)

htcfreek commented:
> Address feedback:
> * click/enter key = open regedit
> * Copy value name / Copy key path = ctrl+c
> * Copy value data = ctrl+shift+c 
> 
> * Key name is copy as `Path + \\ + KeyName `
> 
> @htcfreek
> When I remove the run as admin, I can't longer open `regedit.exe` on my computer with only one user admin account
> 
> ![2021-01-11 21_02_41-Error on open Registry Editor](https://user-images.githubusercontent.com/17874713/104232527-c12cfd00-5450-11eb-8bd7-f83f8b082ed7.png)
> 

I thought regedit can be open without admin rights. 🤔 
Can you please try what happened if the following regkeys are set:
```
Computer\HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\Policies\System" 
- EnableLua=0
- ConsentPromptBehaviorAdmin=0
```
You have to restart after changing the keys.

htcfreek commented on src/modules/launcher/Plugins/Microsoft.Plugin.Registry/Properties/Resources.resx:
```diff
@@ -0,0 +1,174 @@
+    <value>Registry Plugin</value>
+  </data>
+  <data name="PluginDescription" xml:space="preserve">
+    <value>Search inside the Windows Registry</value>
```
We should add a comment that "Windows Registry" as a name. "Windows" shouldn't be translated as in "opening window of my car".

crutkas replied:
Comment should be "this built into Windows the OS.  translate accordingly, https://docs.microsoft.com/de-de/troubleshoot/windows-server/performance/windows-registry-advanced-users is an example of it translated in German"

htcfreek commented on src/modules/launcher/Plugins/Microsoft.Plugin.Registry/Properties/Resources.resx:
```diff
@@ -0,0 +1,174 @@
+    <value>Copy value name</value>
+  </data>
+  <data name="OpenKeyInRegistryEditor" xml:space="preserve">
+    <value>Open key in registry editor</value>
```
We should add a comment that "Registry Editor" is an app name.

Same we need for other occurrences.

htcfreek commented on src/modules/launcher/Plugins/Microsoft.Plugin.Registry/Properties/Resources.resx:
```diff
@@ -0,0 +1,174 @@
+    <value>Error on open Registry Editor</value>
+  </data>
+  <data name="SubKeys" xml:space="preserve">
+    <value>Sub-keys:</value>
```
I would suggest "Subkeys" to make it better understandable for the loc system.

crutkas replied:
Term should mimic what is on docs.microsoft.com.  happy to track this down if i have more context where this is used.  in comment add the URL and let the loc team do appropriate translation as we provided some way finding

crutkas replied:
https://docs.microsoft.com/en-us/windows/win32/sysinfo/enumerating-registry-subkeys

Subkey is correct

htcfreek replied:
> Term should mimic what is on docs.microsoft.com.  happy to track this down if i have more context where this is used.  in comment add the URL and let the loc team do appropriate translation as we provided some way finding

It's used in the subtitle of the results when showing the number of subkeys in a key. 
`Subkeys: 9`

htcfreek reviewed (changes requested):
Please add some comments to the loc file. (cc: @crutkas)


We should decide if we write:
- registry editor
- Windows Registry editor
- Windows Registry Editor
- (other versions of the phrase)

The phrase should be in line with the phrase for "Windows Registry".

htcfreek commented:
@TobiasSekan 
I think the plugin should follow the new naming convention when we merge it. (Issue #9003.)

`Microsoft.Plugin.Registry` - > `Microsoft.PowerToys.Run.Plugin.Registry`

crutkas commented on src/modules/launcher/Plugins/Microsoft.Plugin.Registry/Properties/Resources.resx:
```diff
@@ -0,0 +1,174 @@
+    <value>Search inside the Windows Registry</value>
+  </data>
+  <data name="CopyKeyPath" xml:space="preserve">
+    <value>Copy key path</value>
```
This should be "Key name (path)"
"comment, "Here is how 'The maximum size of a key name is 255 characters.' in german 'Die maximale Größe eines Schlüssel namens beträgt 255 Zeichen.'  https://docs.microsoft.com/en-us/troubleshoot/windows-server/performance/windows-registry-advanced-users to show how the term "key name" should be translated

crutkas commented on src/modules/launcher/Plugins/Microsoft.Plugin.Registry/Properties/Resources.resx:
```diff
@@ -0,0 +1,174 @@
+    <value>You do not have enough rights to open the Windows registry editor</value>
+  </data>
+  <data name="OpenInRegistryEditorAccessExceptionTitle" xml:space="preserve">
+    <value>Error on open Registry Editor</value>
```
lower case registry editor

crutkas commented on src/modules/launcher/Plugins/Microsoft.Plugin.Registry/Properties/Resources.resx:
```diff
@@ -0,0 +1,174 @@
+    <value>Registry key</value>
+  </data>
+  <data name="RegistryValue" xml:space="preserve">
+    <value>Registry Value</value>
```
lower case value

crutkas commented on src/modules/launcher/Plugins/Microsoft.Plugin.Registry/Properties/Resources.resx:
```diff
@@ -0,0 +1,174 @@
+    <value>Registry Value</value>
+  </data>
+  <data name="Type" xml:space="preserve">
+    <value>Type:</value>
```
Comment: See https://docs.microsoft.com/en-us/windows/win32/sysinfo/registry-value-types for proper context of how to translate 'type'

crutkas commented on src/modules/launcher/Plugins/Microsoft.Plugin.Registry/Properties/Resources.resx:
```diff
@@ -0,0 +1,174 @@
+    <value>Type:</value>
+  </data>
+  <data name="Value" xml:space="preserve">
+    <value>Value:</value>
```
Add comment: Comment: See https://docs.microsoft.com/en-us/windows/win32/sysinfo/registry-value-types for proper context of how to translate 'value'

TobiasSekan commented:
@htcfreek my values was
```
Computer\HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\Policies\System" 
- EnableLua=1
- ConsentPromptBehaviorAdmin=5
```

TobiasSekan commented:
@htcfreek When I changed it to 
```
Computer\HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\Policies\System" 
- EnableLua=0
- ConsentPromptBehaviorAdmin=0
```

it will work both (start it as admin and start normal)

htcfreek commented:
> @htcfreek my values was
> ```
> Computer\HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\Policies\System" 
> - EnableLua=1
> - ConsentPromptBehaviorAdmin=5
> ```
> 

This are the default settings (UAC=enabled).

htcfreek commented:
@TobiasSekan 
Why do you add this RunAsAdmin switch in OpenKeyInReg function?

TobiasSekan commented:
> 
> 
> @TobiasSekan
> Why do you add this RunAsAdmin switch in OpenKeyInReg function?

This was for testing, should i remove it?

htcfreek commented:
> > 
> > 
> > @TobiasSekan
> > Why do you add this RunAsAdmin switch in OpenKeyInReg function?
> 
> This was for testing, should i remove it?

Imo code should be simple and free from unnecessary code.

htcfreek commented:
> @TobiasSekan 
> I think the plugin should follow the new naming convention when we merge it. (Issue #9003.)
> 
> `Microsoft.Plugin.Registry` - > `Microsoft.PowerToys.Run.Plugin.Registry`

And please do the necessary changes on the naming.

TobiasSekan commented:
@htcfreek I have rename the projects (Project+UnitTest) and i have  remove the test code

htcfreek reviewed (approved):
Now I think we are ready for the final review, adding to wxs and the merging it.

cc: @enricogior, @crutkas

crutkas commented:
@enricogior, you want to do first pass and i'll take a look shortly.

htcfreek commented:
@TobiasSekan 
Devdocs are out of date.

TobiasSekan commented:
@htcfreek
I have update the documentation 

and have make a few missing name changes
`Microsoft.Plugin.Registry - > Microsoft.PowerToys.Run.Plugin.Registry`

htcfreek commented:
> @htcfreek
> I have update the documentation 
> 
> and have make a few missing name changes
> `Microsoft.Plugin.Registry - > Microsoft.PowerToys.Run.Plugin.Registry`

Do we have now both namespaces used? There are some `Microsoft.Plugin.Registry` left, if see this correctly. But I am not so familiar with this.

TobiasSekan commented:
> 
> 
> > @htcfreek
> > I have update the documentation
> > and have make a few missing name changes
> > `Microsoft.Plugin.Registry - > Microsoft.PowerToys.Run.Plugin.Registry`
> 
> Do we have now both namespaces used? There are some `Microsoft.Plugin.Registry` namespaces left, if see this correctly. But I am not so familiar with this.

You are right, I haven't not change the namesapace too
So now we have only `Microsoft.PowerToys.Run.Plugin.Registry`

htcfreek commented:
@enricogior 
Do you think we can get this in for 0.31?

htcfreek commented:
> @enricogior, you want to do first pass and i'll take a look shortly.

Ping: @enricogior

Jay-o-Way commented:
> @TobiasSekan
> 
> Here's a light and dark theme icon that might be suitable. It somewhat represents the regedit.exe app icon.
> 
> ![registry light](https://user-images.githubusercontent.com/9866362/98598130-4d4c6800-22da-11eb-834e-7de3696b1212.png)
> ![registry dark](https://user-images.githubusercontent.com/9866362/98598135-4de4fe80-22da-11eb-8922-6cded19b8fa2.png)

That's basically `E74C` "OEM" in MDL2 chars.
Why not use the icon directly from `regedit.exe`?

htcfreek commented:
> > @TobiasSekan
> > 
> > Here's a light and dark theme icon that might be suitable. It somewhat represents the regedit.exe app icon.
> > 
> > ![registry light](https://user-images.githubusercontent.com/9866362/98598130-4d4c6800-22da-11eb-834e-7de3696b1212.png)
> > ![registry dark](https://user-images.githubusercontent.com/9866362/98598135-4de4fe80-22da-11eb-8922-6cded19b8fa2.png)
> 
> That's just `E74C` "OEM" in MDL2 font.
> Why not use the icon directly from `regedit.exe`?

Because we use those font like icons and not the colored ones for plugins.

enricogior commented on .pipelines/pipeline.user.windows.yml:
```diff
@@ -118,6 +118,10 @@ build:
             - 'modules\launcher\Plugins\Microsoft.Plugin.Program\Wox.Plugin.dll'
             - 'modules\launcher\Plugins\Microsoft.Plugin.Program\Telemetry.dll'
+            - 'modules\launcher\Plugins\Microsoft.PowerToys.Run.Plugin.Registry\Microsoft.PowerToys.Run.Plugin.Registry.dll'
+            - 'modules\launcher\Plugins\Microsoft.PowerToys.Run.Plugin.Registry\Wox.Infrastructure.dll'
+            - 'modules\launcher\Plugins\Microsoft.PowerToys.Run.Plugin.Registry\Wox.Plugin.dll'
```
Why are these files necessary?
All other plugins don't install them in their folder, they are already installed in the root folder for PT Run.

enricogior replied:
```suggestion
```

enricogior commented on .pipelines/pipeline.user.windows.yml:
```diff
@@ -118,6 +118,10 @@ build:
+            - 'modules\launcher\Plugins\Microsoft.PowerToys.Run.Plugin.Registry\Microsoft.PowerToys.Run.Plugin.Registry.dll'
+            - 'modules\launcher\Plugins\Microsoft.PowerToys.Run.Plugin.Registry\Wox.Infrastructure.dll'
+            - 'modules\launcher\Plugins\Microsoft.PowerToys.Run.Plugin.Registry\Wox.Plugin.dll'
+            - 'modules\launcher\Plugins\Microsoft.PowerToys.Run.Plugin.Registry\Telemetry.dll'
```
This is the telemetry dll for C++ projects, the other plugins use `ManagedTelemetry.dll`

enricogior replied:
```suggestion
            - 'modules\launcher\Plugins\Microsoft.PowerToys.Run.Plugin.Registry\ManagedTelemetry.dll'
```

enricogior commented on .pipelines/pipeline.user.windows.yml:
```diff
@@ -118,6 +118,10 @@ build:
             - 'modules\launcher\Plugins\Microsoft.Plugin.Program\Wox.Plugin.dll'
             - 'modules\launcher\Plugins\Microsoft.Plugin.Program\Telemetry.dll'
+            - 'modules\launcher\Plugins\Microsoft.PowerToys.Run.Plugin.Registry\Microsoft.PowerToys.Run.Plugin.Registry.dll'
+            - 'modules\launcher\Plugins\Microsoft.PowerToys.Run.Plugin.Registry\Wox.Infrastructure.dll'
+            - 'modules\launcher\Plugins\Microsoft.PowerToys.Run.Plugin.Registry\Wox.Plugin.dll'
```
Why are these files necessary?
All other plugins don't install them in their folder, they are already installed in the root folder for PT Run.

enricogior replied:
Actually those files are not even present in the output folder.

TobiasSekan replied:
I was not sure which files are need so I have only taken the same ones as the `pt run service` plugin

enricogior commented:
@TobiasSekan 
do we have an issue to track the work to add the plugin to the installer?

enricogior commented on .pipelines/pipeline.user.windows.yml:
```diff
@@ -118,6 +118,10 @@ build:
+            - 'modules\launcher\Plugins\Microsoft.PowerToys.Run.Plugin.Registry\Microsoft.PowerToys.Run.Plugin.Registry.dll'
+            - 'modules\launcher\Plugins\Microsoft.PowerToys.Run.Plugin.Registry\Wox.Infrastructure.dll'
+            - 'modules\launcher\Plugins\Microsoft.PowerToys.Run.Plugin.Registry\Wox.Plugin.dll'
+            - 'modules\launcher\Plugins\Microsoft.PowerToys.Run.Plugin.Registry\Telemetry.dll'
```
This is the telemetry dll for C++ projects, the other plugins use ManagedTelemetry.dll

enricogior replied:
`Telemetry.dll` is not present in the output folder.

TobiasSekan replied:
I was not sure which files are need so I have only taken the same ones as the pt run service plugin

TobiasSekan commented:
> 
> 
> @TobiasSekan
> do we have an issue to track the work to add the plugin to the installer?

No we need a new Issue

enricogior commented:
@TobiasSekan 
issue created for the installer changes https://github.com/microsoft/PowerToys/issues/9182

enricogior commented:
@TobiasSekan 
I've committed the changes to the files included in `pipeline.user.windows.yml`.
I'll merge the PR and then we can add the plugin to the installer.

enricogior commented:
@crutkas 
the PR is blocked by your change request.

htcfreek commented:
When do we feature close the release 0.31?

crutkas commented:
We need to validate the upgrade path.  I got this working but had to blow away my settings.  Seems good to me, operates how we'd expect.  We will want a good doc page on docs.msft to show how to do searching / shorthands

enricogior commented:
@crutkas 
> We need to validate the upgrade path. I got this working but had to blow away my settings

I didn't run into that problem when running the local build, I'll retest it after we add it to the installer to do a full upgrade test.

htcfreek commented:
> We will want a good doc page on docs.msft to show how to do searching / shorthands

Do we have/need an issue for this?
